src/or/control.c

```diff
@@ -5794,88 +5794,459 @@ control_event_bandwidth_used(uint32_t n_read, uint32_t n_written)
+                                                (edge_connection_t *)exitconn,
+                                                (or_circuit_t *)orcirc);
 }
 
-void control_event_privcount_stream_data_xferred(edge_connection_t *conn, uint64_t amt, int is_outbound) {
-    if(!get_options()->EnablePrivCount || !EVENT_IS_INTERESTING(EVENT_PRIVCOUNT_STREAM_BYTES_TRANSFERRED)) {
-        return;
+/* Is conn an EXIT conn (or BEGINDIR conn), and does it end at this relay? */
+static int
+privcount_is_exit(const edge_connection_t* exitconn,
+                  const or_circuit_t *orcirc)
+{
+  if (!exitconn) {
+    /* Assume missing connections are *not* Exit connections */
+    return 0;
+  }
+
+  const or_circuit_t* oc = privcount_get_const_or_circuit(exitconn, orcirc);
+
+  if (!oc) {
+    /* Assume missing circuits are *not* OR circuits */
+    return 0;
+  }
+
+  /* if the circuit started here, this is our own stream and we can ignore it
+   */
+  if (CIRCUIT_IS_ORIGIN(TO_CIRCUIT(oc))) {
+    return 0;
+  }
+
+  if (!oc->p_chan) {
+    return 0;
+  }
+
+  return TO_CONN(exitconn)->type == CONN_TYPE_EXIT;
+}
+
+/* Should PrivCount ignore events from this connection or circuit?
+ * Directory connections and DNS resolves are considered overhead, so this
+ * function is not suitabe for filtering EVENT_PRIVCOUNT_DNS_RESOLVED events.
+ */
+static int
+privcount_is_overhead(const connection_t* conn, const circuit_t *circ)
+{
+#if PRIVCOUNT_DNS_RESOLVE_IS_OVERHEAD
+  if (privcount_conn_is_dns_resolve(conn)) {
+    return 1;
+  }
+#endif
+
+  /* if the circuit started here, this is our own stream and we can ignore it
+   */
+  if (circ && CIRCUIT_IS_ORIGIN(circ)) {
+    return 1;
+  }
+
+  return privcount_is_dir(conn, circ);
+}
+
+/* Should PrivCount count streams and bytes from this connection or circuit?
+ */
+int
+privcount_is_counted_for_bytes(const edge_connection_t* exitconn,
+                               const or_circuit_t *orcirc)
+{
+  /* Ignore events that are disabled */
+  if (!get_options()->EnablePrivCount) {
+    return 0;
+  }
+
+  if (!EVENT_IS_INTERESTING(EVENT_PRIVCOUNT_STREAM_BYTES_TRANSFERRED) &&
+      !EVENT_IS_INTERESTING(EVENT_PRIVCOUNT_STREAM_ENDED)) {
+    return 0;
+  }
+
+  /* Ignore events with missing connections */
+  if (!exitconn) {
+    return 0;
+  }
+
+  if (privcount_is_overhead(TO_CONN(exitconn), TO_CIRCUIT(orcirc))) {
+    return 0;
+  }
+
+  /* Bytes are only counted at exit connections */
+  return privcount_is_exit(exitconn, orcirc);
+}
+
+/* Should PrivCount ignore cells, circuits, and connections from this circuit?
+ */
+int
+privcount_is_counted_for_cells(const connection_t *conn,
+                               const circuit_t *circ)
+{
+  /* Ignore events that are disabled */
+  if (!get_options()->EnablePrivCount) {
+    return 0;
+  }
+
+  if (!EVENT_IS_INTERESTING(EVENT_PRIVCOUNT_CIRCUIT_ENDED) &&
+      !EVENT_IS_INTERESTING(EVENT_PRIVCOUNT_CONNECTION_ENDED)) {
```

Comment:
The rules are set by the C code: we only increment the counters for events that are active. (We also only increment stream byte counters for exit connections. They are reported as 0 for non-exit connections, I opened #248 to fix this when we do Onion Service statistics.)

I agree this is confusing, and there is a bug in it: the byte counts should be done for the transfer, stream, and circuit events, and the cell counts should be done for the circuit event. I fixed this bug, and a few other errors in these checks, in 812a65bbf.